src/Orc.Controls/Orc.Controls.Shared/Controls/ListTextBox/ListTextBox.cs

```diff
@@ -0,0 +1,252 @@
+//   Copyright (c) 2008 - 2017 WildGums. All rights reserved.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+
+namespace Orc.Controls
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Text;
+    using System.Windows;
+    using System.Windows.Controls;
+    using System.Windows.Input;
+    using System.Windows.Media;
+
+    public class ListTextBox : TextBox
+    {
+        #region Fields
+        private bool _textChangingIsInProgress = false;
+        private int _currentIndex = 0;
+        #endregion
+
+        #region Constructor
+        public ListTextBox()
+        {
+            AddHandler(PreviewMouseLeftButtonDownEvent, new MouseButtonEventHandler(SelectivelyIgnoreMouseButton), true);
+            AddHandler(GotKeyboardFocusEvent, new RoutedEventHandler(SelectAllText), true);
+            AddHandler(MouseDoubleClickEvent, new RoutedEventHandler(SelectAllText), true);
+
+            LostFocus += OnLostFocus;
+        }
+        #endregion
+
+        #region Properties
+        public static readonly DependencyProperty ListOfValuesProperty = DependencyProperty.Register("ListOfValues", typeof(IList<string>),
+            typeof(ListTextBox), new FrameworkPropertyMetadata(null, FrameworkPropertyMetadataOptions.BindsTwoWayByDefault));
+
+        public static readonly DependencyProperty ValueProperty = DependencyProperty.Register("Value", typeof(string),
+            typeof(ListTextBox), new FrameworkPropertyMetadata(null, FrameworkPropertyMetadataOptions.BindsTwoWayByDefault, (sender, e) => ((ListTextBox)sender).OnValueChanged()));
+
+        public IList<string> ListOfValues
+        {
+            get { return (IList<string>)GetValue(ListOfValuesProperty); }
+            set { SetValue(ListOfValuesProperty, value); }
+        }
+
+        public string Value
+        {
+            get { return (string)GetValue(ValueProperty); }
+            set { SetValue(ValueProperty, value); }
+        }
+
+        private bool AllTextSelected
+        {
+            get { return SelectedText == Text; }
+        }
+
+        private bool CaretAtStart
+        {
+            get { return CaretIndex == 0; }
+        }
+
+        private bool CaretAtEnd
+        {
+            get { return CaretIndex == Text.Length; }
+        }
+        #endregion
+
+        #region Events
+        public event EventHandler RightBoundReached;
+        public event EventHandler LeftBoundReached;
+        #endregion
+
+        #region Methods
+        private void OnLostFocus(object sender, RoutedEventArgs e)
+        {
+            SetCurrentValue(ValueProperty, Text);
+            UpdateText();
+        }
+
+        protected override void OnPreviewKeyDown(KeyEventArgs e)
+        {
+            base.OnPreviewKeyDown(e);
+
+            if (e.Key == Key.Right && (CaretAtEnd || CaretAtStart && AllTextSelected))
```

Comment:
CaretAtEnd & CaretAtStart could be expensive. Try setting these in variables and re-use them in the several checks in this method.